### PR TITLE
Framework: data-layer: implement a parameters control in the http() helper.

### PR DESCRIPTION
**Pull request by retrofox**
This patch attempts to improve the process which describes the HTTP request built by the `http()` function. The idea is checking the request parameters controlling when they are needed, type of value, etc.
For instance, `path` and `method` parameters must be defined, `path` should be a string, `method` just could take either `GET` or `POST` values, and so on.

The most important checking is detecting if both `apiVersion` and `apiNamespace` parameters are simultaneously defined. It will avoid reintroducing a bug: More info in the revert PR: https://github.com/Automattic/wp-calypso/pull/11267#issuecomment-278438016

If the request parameteres are ok the `http()` function will return a `WPCOM_HTTP_REQUEST` action type. If not the Redux action returned will be `WPCOM_HTTP_BAD_REQUEST`.

### Testing

```cli
> npm run test-client client/state/data-layer/wpcom-http/test/http.js
```

```
  state
    data-layer
      wpcom-http
        http
          request action
            ✓ should return a valid redux action
            ✓ should return a valid redux action defining apiVersion
            ✓ should return a valid redux action defining apiNamespace
          bad request action
            ✓ should return a bad action when `path` is not defined
            ✓ should return a bad action when `path` is invalid
            ✓ should return a bad action when method is not defined
            ✓ should return a bad action when method is unknown
            ✓ should return a bad action when apiVersion is invalid
            ✓ should return a bad action when apiNamespace is invalid
            ✓ should return a bad action when both apiVersion and apiNamespace are simultaneously defined

  10 passing (760ms)
```

### CODE REVIEW

**Comment by retrofox** [comment]
@retrofox did this not cause the problems with plans earlier? Why are we adding the same code back in?

No. it is not the same code.

> Additionally, it may just be my ignorance, but apiNamespace was a surprise to me. We probably need to do something in terms of documenting the whens and whys of this parameter.

Yes, you're right. There are a lot of remaining work to do in terms of documentation. You could ready a little bit about the `apiNamespace` parameter in the `xhr` handler repository. https://github.com/Automattic/wpcom-xhr-request#api
There there are some examples of how to use this parameter.

> Is it even necessary here?

Absolutely necessary if you need to define the `WP REST API`.

> If it's optional and passed into the query params why add it here

To be consistent with the whole implementation. 

```js
 * @param {string} [apiVersion] specific API version for request
 * @param {Object} [body] JSON-serializable body for POST requests
 * @param {string} method name of HTTP method to use
 * @param {string} path WordPress.com API path with %s and %d placeholders, e.g. /sites/%s
 * @param {Object} [query] key/value pairs for query string
 * @param {FormData} [formData] key/value pairs for POST body, encoded as "multipart/form-data"
 * @param {Object} [onSuccess] Redux action to call when request succeeds
 * @param {Object} [onFailure] Redux action to call when request fails
 * @param {Object} [onProgress] Redux action to call on progress events from an upload
 * @param {Object} [options] extra options to send to the middleware, e.g. retry policy or offline policy
 * @param {Object} [action] default action to call on HTTP events
 * @returns {Object} Redux action describing WordPress.com API HTTP request
```

In this jsdoc `apiVersion`, `body`, `query`, etc... are optional parameters as well, just like `apiNamespace`.

**Comment by retrofox** [comment]
One thing that I suggest is that we should define `apiVersion` as a required parameter, removing its default value (`v1`). In this way, we would be forcing to define the api Version and consequently check that endpoint version is correct.

**Comment by retrofox** [comment]
Set apiVersion parameter as needed: 6059d2b4366422d170333c39fb51b6e112a73ff4

File: client/state/data-layer/wpcom-http/actions.js [hunk 1]
```diff
@@ -30,15 +40,49 @@ export const http = ( {
...
+	}
+
+	// - One of `apiVersion` of `apiNamespace` should be defined
+	// - `apiVersion` and `apiNamespace` cannot be defined simultaneously
+	if (
+		( typeof apiVersion === 'undefined' && typeof apiNamespace === 'undefined' ) ||
+		( typeof apiVersion !== 'undefined' && typeof apiNamespace !== 'undefined' )
+	) {
+		return badRequest();
+	}
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/actions.js]: I'm curious why we've chosen to return a Redux action here. If there's never a time when we want these requests to fly should we not simply abort by throwing an error? fail-early fail-fast? Redux actions would be hard to discover and would indicate that this is something that could happen at runtime and I don't see how we could have these at runtime but not development-time unless we were building an API browser of some kind.

File: client/state/data-layer/wpcom-http/actions.js [hunk 2]
```diff
@@ -30,15 +40,49 @@ export const http = ( {
...
+
+	// - `apiVersion` must have the `v1, v2, ...` shape
+	if ( typeof apiVersion !== 'undefined' && ( /^v\d+/ ).test( apiVersion ) ) {
+		actionProperties.query.apiVersion = apiVersion;
+	}
+
+	// - `apiNamesapce` must have the `wp/v2, wpcom/v2, ...` shape
+	if ( typeof apiNamespace !== 'undefined' && ( /^[a-z]+\/v\d+/ ).test( apiNamespace ) ) {
+		actionProperties.query.apiNamespace = apiNamespace;
+	}
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/actions.js]: could we move these tests into separate functions and run some chain over them?

```js
const noVersionAndNamespaceConflicts = ( { apiNamespace, apiVersion } ) => {
	if ( apiNamespace && apiVersion ) {
		throw new TypeError( 'Cannot simultaneously define `apiNamespace` and `apiVersion` (see wpcom.js/stuff/README.md)' );
	}
}

const apiVersionPattern = /^v\d+/;
const hasViableVersion = ( { apiVersion } ) => {
	if ( ! ( apiVersion && apiVersionPattern.test( apiVersion ) ) {
		throw new TypeError( `Given 'apiVersion' of "${ apiVersion }" doesn't match pattern: v1, v2, etc…` );
	}
}

…

const validators = [
	noVersionAndNamespaceConflicts,
	hasViableVersion,
	hasViableNamespace,
];

…

export const http = request => {
	const {
		apiVersion,
		apiNamespace,
		…
	} = request;

	// validate request and throw on error
	validators.forEach( validator => validator( request ) );

	…
}
```

File: client/state/data-layer/wpcom-http/actions.js [hunk 3]
```diff
@@ -30,15 +40,49 @@ export const http = ( {
...
+
+	return {
+		...actionProperties,
+		type: WPCOM_HTTP_REQUEST,
+		path,
+		method,
+		body,
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/actions.js]: did we not want `formData`?

File: client/state/data-layer/wpcom-http/actions.js [hunk 4]
```diff
@@ -30,15 +40,49 @@ export const http = ( {
...
+	) {
+		return badRequest();
+	}
+
+	// - `apiVersion` must have the `v1, v2, ...` shape
+	if ( typeof apiVersion !== 'undefined' && ( /^v\d+/ ).test( apiVersion ) ) {
+		actionProperties.query.apiVersion = apiVersion;
+	}
+
+	// - `apiNamesapce` must have the `wp/v2, wpcom/v2, ...` shape
```
Comment (gwwar) [inline client/state/data-layer/wpcom-http/actions.js]: @nylen or @retrofox Could we get a bit more clarification on what these namespaces do?

I assume:
- `wp` = WordPress Rest API
- `wpcom` = WordPress.com API

And for anything that has a namespace, any version property is ignored, since that namespace includes a version?

Normally if we are hitting completely different endpoint apis I'd expect 2 client libs, but I'm sure we can clean up the interface so that this is more clear

As @retrofox pointed out earlier it appears that the proxy is only doing a truthy check, which can lead to bugs if you explicitly set a namespace that isn't wp
https://github.com/Automattic/wpcom-proxy-request/blob/master/index.js#L429

File: client/state/data-layer/wpcom-http/actions.js [hunk 5]
```diff
@@ -30,15 +40,49 @@ export const http = ( {
...
+	if (
+		( typeof apiVersion === 'undefined' && typeof apiNamespace === 'undefined' ) ||
+		( typeof apiVersion !== 'undefined' && typeof apiNamespace !== 'undefined' )
+	) {
+		return badRequest();
+	}
+
+	// - `apiVersion` must have the `v1, v2, ...` shape
+	if ( typeof apiVersion !== 'undefined' && ( /^v\d+/ ).test( apiVersion ) ) {
+		actionProperties.query.apiVersion = apiVersion;
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/actions.js]: in the end, let's merge these in in one declarative statement instead of directly mutating the query object - will make for significantly easier code to follow and maintain

**Comment by dmsnell** [comment]
one round of review @retrofox - thanks for working on this!

File: client/state/data-layer/wpcom-http/actions.js (hunk 1 above)
Comment (gwwar) [inline client/state/data-layer/wpcom-http/actions.js]: I think it's fair to gracefully handle malformed requests here, since our callers already have onError handlers. Presumably if we're developing something new we'd notice when it doesn't work.

Additional things to do here would be to add an eslint rule to detect this case so we can't commit the mis-match of version/api namespace, and also update the wpcom.js interface to throw an Error immediately in this case.

File: client/state/data-layer/wpcom-http/actions.js [hunk 6]
```diff
@@ -1,12 +1,21 @@
...
  */
-import { WPCOM_HTTP_REQUEST } from 'state/action-types';
+import {
+	WPCOM_HTTP_REQUEST,
+	WPCOM_HTTP_BAD_REQUEST,
+} from 'state/action-types';
+
+const badRequest = () => ( {
+	type: WPCOM_HTTP_BAD_REQUEST,
+	error: 'request parameters object is not valid'
```
Comment (gwwar) [inline client/state/data-layer/wpcom-http/actions.js]: Can we give more specific errors of why it's bad instead of a generic message?

File: client/state/data-layer/wpcom-http/actions.js (hunk 1 above)
Comment (retrofox) [inline client/state/data-layer/wpcom-http/actions.js]: I've updated the code. I've added warnings when the http request is malformed.

File: client/state/data-layer/wpcom-http/actions.js (hunk 2 above)
Comment (retrofox) [inline client/state/data-layer/wpcom-http/actions.js]: for sure. @dmsnell I've updated the tests just now. I didn't ready your comments before to push.

**Comment by retrofox** [comment]
> Well this is pretty odd, I'm seeing an infinite loader for only the personal plan

Yes, that's right. I was completely aware of this and I've fixed the bug. the apiVersion RegExp wasn't allowing the `v1.4` shape.

File: client/state/data-layer/wpcom-http/actions.js [hunk 7]
```diff
@@ -30,15 +49,66 @@ export const http = ( {
...
+	}
+
+	// optional parameters
+	if ( typeof formData !== 'undefined' ) {
+		actionProperties.formData = [];
+	}
+
+	return {
+		...actionProperties,
+		...options,
```
Comment (retrofox) [inline client/state/data-layer/wpcom-http/actions.js]: @dmsnell I'm pretty sure we should destructuring this object here, which we aren't doing in prod. I've added a test which passes a random option into the params object. Could you confirm this?

File: client/state/data-layer/wpcom-http/actions.js (hunk 3 above)
Comment (retrofox) [inline client/state/data-layer/wpcom-http/actions.js]: Updated. I was working on this ;-)

**Comment by retrofox** [comment]
@nylen could you take a look to these tests please? https://github.com/Automattic/wp-calypso/pull/11280/commits/2965a6f7cf932de0a5b5b28a290191cc7aaac30d
Let me know if we should update the regexp or add more test.

File: client/state/data-layer/wpcom-http/actions.js (hunk 6 above)
Comment (retrofox) [inline client/state/data-layer/wpcom-http/actions.js]: yes, done.

**Comment by retrofox** [comment]
> Additional things to do here would be to add an eslint rule to detect this case

100% agree

File: client/state/data-layer/wpcom-http/actions.js (hunk 1 above)
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/actions.js]: the point is, why would we create a Redux action for something that should never exist?

this is a 100% failure and will only happen at runtime if we mess up the code. it doesn't make any sense to me to treat it as if it's valid.

_why_ do we want to create the Redux actions in the first place? _why_ would we not want to entirely abort on invalid code?

File: client/state/data-layer/wpcom-http/actions.js (hunk 7 above)
Comment (retrofox) [inline client/state/data-layer/wpcom-http/actions.js]:
```js
const action = http( {
 	apiVersion: 'v2',
 	foo: 'bar',
 	method: 'GET',
 	path: '/path/to/endpoint',
 } );
 
 expect( action ).to.eql( {
 	type: WPCOM_HTTP_REQUEST,
 	body: {},
 	method: 'GET',
 	foo: 'bar',
 	path: '/path/to/endpoint',
 	query: {
 		apiVersion: 'v2',
 	},
 	onSuccess: null,
 	onFailure: null,
 	onProgress: null,
 } );
```

File: client/state/data-layer/wpcom-http/actions.js (hunk 1 above)
Comment (gwwar) [inline client/state/data-layer/wpcom-http/actions.js]: Would we never programmatically craft an http() request? If not then sure, we can stick with static analysis or explicit Error thowing.

Comment (retrofox) [inline client/state/data-layer/wpcom-http/actions.js]: > why do we want to create the Redux actions in the first place? why would we not want to entirely abort on invalid code?

I think you're right. We should abort and throw an error. If not we will silence/allowing something that shouldn't be allowed.

File: client/state/data-layer/wpcom-http/actions.js (hunk 4 above)
Comment (retrofox) [inline client/state/data-layer/wpcom-http/actions.js]: `wp` - `WP REST API` -> default endpoints defined in the wpapi.org plugin.
discovery endpoint: https://public-api.wordpress.com/wp/v2

`wpcom` - `WP REST API` -> custom endpoints defined in .com
discovery endpoint: https://public-api.wordpress.com/wpcom/v2

If I'm not wrong ...

**Comment by dmsnell** [comment]
> I think it's fair to gracefully handle malformed requests here, since our callers already have onError handlers.

@gwwar this isn't exactly the same. the `onFailure` responder action is there for network failures and for when the API returns error responses (such as a `501` or `404`). They aren't there for handling coding errors. It's really something that should never surface to a user.

File: client/state/data-layer/wpcom-http/actions.js (hunk 7 above)
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/actions.js]: @retrofox `options` isn't for destructuring. it's not a part of the HTTP request proper but rather an object into which we can pass policies describing how the system should handle the HTTP request. if we spread it out then that will break the system

File: client/state/data-layer/wpcom-http/actions.js [hunk 8]
```diff
@@ -1,12 +1,30 @@
...
  */
-import { WPCOM_HTTP_REQUEST } from 'state/action-types';
+import {
+	WPCOM_HTTP_REQUEST,
+	WPCOM_HTTP_BAD_REQUEST,
+} from 'state/action-types';
+import warn from 'lib/warn';
+
+export const apiVersionRegex = new RegExp( /^v(?:(\d+)\.)?(\d+)$/ );
+export const apiNamespaceRegex = new RegExp( /^(?:([a-z]+)\/)?v(\d+)$/ );
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/actions.js]: there's no need to write `new Regexp` here if you don't want to. It doesn't offer any advantage over creating a RegExp literal `/something/`

File: client/state/data-layer/wpcom-http/actions.js [hunk 9]
```diff
@@ -17,10 +35,11 @@ import { WPCOM_HTTP_REQUEST } from 'state/action-types';
  * @param {Object} [onProgress] Redux action to call on progress events from an upload
  * @param {Object} [options] extra options to send to the middleware, e.g. retry policy or offline policy
  * @param {Object} [action] default action to call on HTTP events
- * @returns {Object} Redux action describing WordPress.com API HTTP request
+ * @return {Object} Redux action describing WordPress.com API HTTP request
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/actions.js]: why did you change this?

File: client/state/data-layer/wpcom/plans/index.js [hunk 10]
```diff
@@ -23,15 +26,19 @@ import {
...
 		onFailure: action,
-	} ) );
+	} );
+
+	dispatch( httpAction );
 
-	return next( action );
+	if ( httpAction.type !== WPCOM_HTTP_BAD_REQUEST ) {
+		next( action );
+	}
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/plans/index.js]: this code here is one of the reasons I feel very strongly against creating a Redux action for bad requests. A request which is always invalid and we can tell at compile-time should never be checked for in other parts of the code otherwise we'll end up with hundreds of redundant checks that aren't needed and pollute other code. Let's kill the error at the source instead of propagating it around the app.

**Comment by retrofox** [comment]
> > I think it's fair to gracefully handle malformed requests here, since our callers already have onError handlers.

> @gwwar this isn't exactly the same. the onFailure responder action is there for network failures and for when the API returns error responses (such as a 501 or 404). They aren't there for handling coding errors. It's really something that should never surface to a user.

Yup, testing the plans stuff I forced an error passing a bad apiVersion. Then I had to add a condition checking if it was a `WPCOM_HTTP_BAD_REQUEST` action to aviod the `next( action )` execution. Here is the commit: https://github.com/Automattic/wp-calypso/pull/11280/commits/91e0727424819b8bd0d09c34995e1686a455c654

**Comment by gwwar** [comment]
A more common case would be something where path would be invalid like `sites/undefined/bar` if we're sloppy with our variables in our string interpolation. Should we just let it throw an error in that case?

File: client/state/data-layer/wpcom/plans/index.js (hunk 10 above)
Comment (retrofox) [inline client/state/data-layer/wpcom/plans/index.js]: image

Comment (retrofox) [inline client/state/data-layer/wpcom/plans/index.js]: https://github.com/Automattic/wp-calypso/pull/11280#issuecomment-278808540

File: client/state/data-layer/wpcom-http/actions.js (hunk 4 above)
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/actions.js]: it's not that important that @gwwar and I know what these do here. what matters is that these things need to be obvious and discoverable inside of the code because we're not the people who will be using it the most. if _we're_ asking the question then we can expect many other people will be asking the same question but they won't have this PR discussion to look over. let's get it clear in some documentation and make it obvious for people wanting to use this code.

File: client/state/data-layer/wpcom-http/actions.js (hunk 7 above)
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/actions.js]: @retrofox what is `foo` doing there? it shouldn't be in there. what were you trying to communicate with your code snippet?

**Comment by retrofox** [comment]
> A more common case would be something where path would be invalid like sites/undefined/bar if we're sloppy with our variables in our string interpolation. Should we just let it throw an error in that case?

Well ... we could do it but to achieve that we should know in advance the structure of all endpoints. it's quite easy to know for the `WP REST API` through of the discover endpoints.
But I think that correctly defining these values is still a responsibility of the developer who is implementing the layer.

**Comment by gwwar** [comment]
Thinking about this further, we can still run into the same issue when calling wpcom.js directly right @retrofox? Folks who are working on the wpcom.js lib, also won't think to update rules here, if say we use a new namespace format. Let's move these validation checks to wpcom.js and safeguard with an eslint rule in Calypso.

File: client/state/data-layer/wpcom-http/actions.js (hunk 7 above)
Comment (retrofox) [inline client/state/data-layer/wpcom-http/actions.js]: yup, my bad. I got confused on this. Reverting this change.

**Comment by gwwar** [comment]
> But I think that correctly defining these values is still a responsibility of the developer who is implementing the layer.

@retrofox the case I was thinking of is say when we craft a request without checking if say siteId is defined.

```js
const path = `sites/${ siteId }/path`;
```

But you make a fair point 👍 I suppose we'd want to be handling any cases that slip through code review via monitoring anyway.

File: client/state/data-layer/wpcom-http/actions.js (hunk 9 above)
Comment (retrofox) [inline client/state/data-layer/wpcom-http/actions.js]: I used to use `@return` instead of `@returns`. In fact, the number of `@return` is clearly greater than `@returns` in the base code of calypso. And it seems that both are accepted.

File: client/state/data-layer/wpcom-http/actions.js (hunk 4 above)
Comment (gwwar) [inline client/state/data-layer/wpcom-http/actions.js]: >  let's get it clear in some documentation and make it obvious for people wanting to use this code.

Exactly this 👍 Though I'm sure you had some PRs planned for this @retrofox and @nylen ?

Perhaps we should put this PR on hold until we get some better documentation PRs in wpcom.js and have it be much more obvious and strict when we give it bad arguments? I'd be happy to take a look and help test.

Does something like this order make sense?
- wpcom.js documentation PR
- wpcom.js validation + version bump PR
- bump wpcom.js in Calypso, with full regression testing.
- update wpcom-http to accept apiNamespace PR
- (optional) eslint rule / calypso integration

**Comment by dmsnell** [comment]
> @retrofox the case I was thinking of is say when we craft a request without checking if say siteId is defined.

I'm less worried about these cases @gwwar too because they will end up flagging `404s` and we will see that in the console. definitely _not_ opposed to catching them early in development but I would imagine it's easier to make this mistake at runtime assuming that some site id is set when it's not than actively writing an invalid combination of parameters.

**Comment by retrofox** [comment]
hi @dmsnell, I've updated and rebased the code. Please feel free to make a new round when you have the chance. Thanks Dennis.

File: client/state/data-layer/wpcom-http/actions.js [hunk 11]
```diff
@@ -3,10 +3,75 @@
...
+	if ( apiNamespace && apiVersion ) {
+		throw new TypeError(
+			'Cannot simultaneously define `apiNamespace` and `apiVersion` (see wpcom.js/README.md)'
+		);
+	}
+};
+
+const hasViableVersion = ( { apiVersion, apiNamespace } ) => {
+	if ( ! apiVersion && apiNamespace ) {
+		return;
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/actions.js]: this return statement is awfully ambiguous. the function has a predicate name `has_` but returns `undefined` on every invocation. does that mean is _does_ have a viable version? `undefined` is falsely so if we were to see `if ( hasViableVersion( … ) )` then that conditional would always return false.

what if we simply returned `true` in all cases where we don't `throw`

File: client/state/data-layer/wpcom-http/actions.js (hunk 9 above)
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/actions.js]: okay, let's just leave it in then as-is if you didn't have a reason it was wrong. if we're going to make a purely stylistic change we can do it all at once in one PR which doesn't change behaviors.

File: client/state/data-layer/wpcom-http/actions.js [hunk 12]
```diff
@@ -17,28 +82,57 @@ import { WPCOM_HTTP_REQUEST } from 'state/action-types';
...
+
+	// formData is optional
+	if ( formData ) {
+		actionProperties.formData = FormData;
+	}
+
+	// optional parameters
+	if ( typeof formData !== 'undefined' ) {
+		actionProperties.formData = [];
+	}
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/actions.js]: if we've already validated our data why then do we need all these conditionals? it seems like we should just combine them into the action as if they are valid…

```js
return {
	query: { ...query, apiVersion, apiNamespace }
}
```

File: client/state/data-layer/wpcom-http/actions.js [hunk 13]
```diff
@@ -17,28 +82,57 @@ import { WPCOM_HTTP_REQUEST } from 'state/action-types';
...
+	validators.forEach( validator => validator( request ) );
+
+	const actionProperties = {
+		body,
+		onSuccess: onSuccess || action,
+		onFailure: onFailure || action,
+		onProgress: onProgress || action,
+		query: { ...query },
+		options,
+	};
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/actions.js]: this object seems unnecessary - why not simply put these properties directly in the return statement below?

File: client/state/data-layer/wpcom-http/test/http.js [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+			expect( apiVersionPattern.test( 'v1' ) ).to.be.true;
+		} );
+
+		it( 'should not accept `v1.` as a valid version', () => {
+			expect( apiVersionPattern.test( 'v1.' ) ).to.be.false;
+		} );
+
+		it( 'should accept `v1.3` as a valid version', () => {
+			expect( apiVersionPattern.test( 'v1.3' ) ).to.be.true;
+		} );
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/test/http.js]: maybe we could make a list of valid versions and invalid versions and then run the test over that list. I think it would remove a lot of redundant code (which isn't actually helping us understand the functionality better) and leave us with a cleaner/simpler way of assessing what _should_ and _shouldn't_ be acceptable values.

```js
describe( 'apiVersion RegExp pattern', () => {
	const validVersions = [ 'v1', 'v1.3' ];
	const invalidVersions = [ '1', 'v1.' ];

	it( 'should pass valid versions', () => {
		validVersions.forEach( v => expect( apiVersionPattern.test( v ) ).to.be.true );
	} );

	it( 'should fail invalid versions', () => {
		invalidVersions.forEach( v => expect( apiVersionPattern.test( v ) ).to.be.false );
	} );
} );
```

now we can more easily add test cases to the tests while keeping the suite small.

> note! these are spot-check tests which catch a very limited set of inputs. we need to be aware of the limitation here

File: client/state/data-layer/wpcom-http/test/http.js [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+			expect( apiNamespacePattern.test( 'wpcom/v1' ) ).to.be.true;
+		} );
+
+		it( 'should not accept `wpcom/` as a valid namespace', () => {
+			expect( apiNamespacePattern.test( 'wpcom/' ) ).to.be.false;
+		} );
+
+		it( 'should not accept `wpcom/1` as a valid namespace', () => {
+			expect( apiNamespacePattern.test( 'wpcom/1' ) ).to.be.false;
+		} );
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/test/http.js]: see my comment above on valid/invalid test cases

File: client/state/data-layer/wpcom-http/test/http.js [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+				method: 'GET',
+				path: '/path/to/endpoint',
+				query: {
+					apiVersion: 'v2'
+				},
+				options: {},
+				onSuccess: null,
+				onFailure: null,
+				onProgress: null,
+			} );
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/test/http.js]: if the test is checking for a valid `apiVersion` then let's not add in a bunch of other tests to it. this is baking fragility into the test by preventing the `http()` function from changing - it's an anti-unit test

instead, we can make the test focus on the testable behavior…

```js
expect( action.query.apiVersion ).to.equal( 'v2' );
```

…and leave everything else out of it

File: client/state/data-layer/wpcom-http/test/http.js [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+	} );
+
+	describe( 'bad request action', () => {
+		it( 'should throw Error when `apiVersion` and `apiNamespace` are not defined.', () => {
+			const boundRequest = http.bind( http, {
+				path: '/path/to/endpoint',
+				method: 'GET',
+			} );
+
+			expect( boundRequest ).to.throw( 'One of `apiNamespace` and `apiVersion` must be defined (see wpcom.js/README.md)' );
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/test/http.js]: are we really concerned about testing the specific error messages our validators throw or simply _that_ they throw.

File: client/state/data-layer/wpcom-http/test/http.js [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+			} );
+		} );
+	} );
+
+	describe( 'bad request action', () => {
+		it( 'should throw Error when `apiVersion` and `apiNamespace` are not defined.', () => {
+			const boundRequest = http.bind( http, {
+				path: '/path/to/endpoint',
+				method: 'GET',
+			} );
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-http/test/http.js]: this is a bit odd to see here - as if it were working around a problem that didn't exist (simply because of the way `bind` is given `http` when in fact we don't care about that.

how about just creating a fat-arrow function?

```js
const withoutVersioning = () => http( { … } );

expect( withoutVersioning ).to.throw( TypeError );
```

**Comment by dmsnell** [comment]
@retrofox thanks for the further updates. I think that the validators belong somewhere besides `actions.js` because they aren't actions in any sense. it would be reasonable to have a new subfolder (`validators`) with its own README describing what they are and how people can create new ones or modify the old ones. it would be lovely to have this in a walk-through style guide instead of simply listing the functions/API. that would help new contributors to _understand_ the whys and hows of the system.

the tests are also fairly hefty and I don't think they need to be. there's much more in there than unit tests and they actually test far more than their names lead on.

File: client/state/data-layer/wpcom-http/actions.js (hunk 9 above)
Comment (retrofox) [inline client/state/data-layer/wpcom-http/actions.js]: Let me revert it.

**Comment by dmsnell** [comment]
@retrofox what's the status on this? I feel like we might be able to just let `wpcom.js` handle the validation, possibly moving the calls there. or, if we _need_ it in Calypso, that we could possibly do it in a slightly different way. does it need an update? should we close and abandon this?

**Comment by retrofox** [comment]
> @retrofox what's the status on this? I feel like we might be able to just let wpcom.js handle the validation, possibly moving the calls there. or, if we need it in Calypso, that we could possibly do it in a slightly different way. does it need an update? should we close and abandon this?

if we are going to handle this in wpcom.js then yes, let's close this PR. We needed a more immediate solution.
